dbt/tests/data_tests/generic_tests/expect_column_fraction_with_condition.sql

```diff
@@ -1,20 +1,30 @@
 {% test expect_column_fraction_with_condition(
     model,
     column_name,
-    column_condition,
-    expected_fraction,
-    tol=0.02
+    row_condition,
+    min_fraction,
+    max_fraction
 ) %}
 
+WITH fraction_check AS (
+    SELECT
+        SUM(CASE WHEN {{ row_condition }} THEN {{ column_name }} ELSE 0 END) * 1.0 / SUM({{ column_name }}) AS actual_fraction,
+        SUM(CASE WHEN {{ row_condition }} THEN {{ column_name }} ELSE 0 END) AS condition_sum,
+        SUM({{ column_name }}) AS total_sum,
+    FROM {{ model }}
+)
+
```

Comment:
In order to build useful failure messages I've found it really helpful to make these little temporary tables (aka "Common Table Expressions" or CTEs) and then select from them.